lib/vm/src/instance/ref.rs

```diff
@@ -155,3 +155,75 @@ impl InstanceRef {
         (&mut *ptr).as_mut()
     }
 }
+
+/// A weak instance ref. This type does not keep the underlying `Instance` alive
+/// but can be converted into a full `InstanceRef` if the underlying `Instance` hasn't
+/// been deallocated.
+#[derive(Debug, Clone)]
+pub struct WeakInstanceRef(Weak<InstanceInner>);
+
+impl PartialEq for WeakInstanceRef {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.ptr_eq(&other.0)
+    }
+}
+
+impl WeakInstanceRef {
+    /// Try to convert into a strong, `InstanceRef`.
+    pub fn upgrade(&self) -> Option<InstanceRef> {
+        let inner = self.0.upgrade()?;
+        Some(InstanceRef(inner))
+    }
+}
+
+impl MemoryUsage for WeakInstanceRef {
+    fn size_of_val(&self, tracker: &mut dyn MemoryUsageTracker) -> usize {
+        mem::size_of_val(self)
+            + if let Some(ir) = self.upgrade() {
+                ir.size_of_val(tracker)
+            } else {
+                0
+            }
+    }
+}
+
+/// An `InstanceRef` that may or may not be keeping the `Instance` alive.
+///
+/// This type is useful for types that conditionally must keep / not keep the
+/// underlying `Instance` alive. For example, to prevent cycles in `WasmerEnv`s.
+#[derive(Debug, Clone, PartialEq, MemoryUsage)]
+pub enum WeakOrStrongInstanceRef {
+    /// A weak instance ref.
+    Weak(WeakInstanceRef),
+    /// A strong instance ref.
+    Strong(InstanceRef),
+}
+
+impl WeakOrStrongInstanceRef {
+    /// Get a strong `InstanceRef`. A return Value of `None` means that the
+    /// `InstanceRef` has been freed and cannot be accessed.
+    pub fn get_strong(&self) -> Option<InstanceRef> {
```

Comment:
`get_strong` is always used as `WeakOrStrongInstanceRef::Strong(v.get_strong().unwrap())`.

Because of this, I think it might be better to have a `try_into_strong()` method that returns a `Option<WeakOrStrongInstanceRef>` and will also simplify this code.

Reply:
Nice observation about how it's used, we could use `TryInto` or something for these conversions